subprojects/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishHttpIntegTest.groovy

```diff
@@ -384,4 +385,43 @@ credentials {
         module.jarFile.assertExists()
         module.ivyFile.assertDoesNotExist()
     }
+
+    def "retries artifact upload for transient network error"() {
+        given:
+        server.start()
+        settingsFile << 'rootProject.name = "publish"'
+        buildFile << """
+            apply plugin: 'java'
+            apply plugin: 'ivy-publish'
+
+            version = '2'
+            group = 'org.gradle'
+
+            publishing {
+                repositories {
+                    ivy { url "${ivyHttpRepo.uri}" }
+                }
+                publications {
+                    ivy(IvyPublication) {
+                        from components.java
+                    }
+                }
+            }
+        """
+
+        and:
+        module.jar.expectPutBroken()
```

Comment:
We should probably also have a test that it provides a consumable error when all the retries fail (unless you think this is covered properly elsewhere).

We might also think about having this test produce multiple failures (less than max attempts) to validate that we can recover after multiple retries.

Reply:
The existing test `IvyPublishHttpIntegTest.reports failure publishing to HTTP repository` actually test the failure case, since it uses `server.addBroken()`. This causes the server to respond with a 500 error, which triggers a retry.

Reply:
I added some additional coverage in 94e9fb2b14bd5aa9e120b2746d3ed4f92cd331fc.
- Test multiple retries in integration test, as well as test retry of POM/Ivy/Module files.
- Test success-after-retry in unit test